File: scripts/fix_meanings.py
"""
Fix root meanings and generate specific bab verb meanings using Claude API.
Run: python3 scripts/fix_meanings.py
Requires: ANTHROPIC_API_KEY env var
"""

import json
import os
import time
import sys
from pathlib import Path

import anthropic

DATA_PATH = Path(__file__).parent.parent / "public" / "data" / "verbsData.json"
BATCH_SIZE = 10  # roots per API call

client = anthropic.Anthropic(api_key=os.environ.get("ANTHROPIC_API_KEY"))


def looks_bad(meaning: str) -> bool:
    """Detect transliterated or obviously wrong meanings."""
    if not meaning or len(meaning.strip()) == 0:
        return True
    m = meaning.lower().strip()
    # Remove "to " prefix for analysis
    core = m.removeprefix("to ").strip()
    # If core is a single word that looks like a transliteration (no real English meaning)
    bad_signals = [
        len(core.split()) == 1 and len(core) <= 4 and core.isalpha(),  # short single word
        core in {"sayh", "jia", "shea", "rod", "boa", "zako", "neil", "basl",
                 "being", "came", "msds", "gurrr", "nbz", "sjr", "sabo", "atto",
                 "arw", "dso", "msu", "lvo", "hafo", "sagho", "sjoo", "dahu",
                 "tahu", "otho", "fazo", "hout", "qadda", "nazzg", "zjo", "asho",
                 "waka", "fazz", "safwa", "mona", "aso", "noob", "dom", "cor",
                 "sam", "ken", "sol", "sram", "cbcb", "cbb", "nx", "rex", "bor",
                 "lafs", "haig", "mer", "fata", "boa", "yum", "ren", "mello",
                 "glo", "nbat", "ism", "sjoo", "psst", "tahu", "watta", "gus",
                 "ptk", "lvf", "sjr", "dso", "msu", "bahl", "dhak", "dhaha",
                 "sfg", "spg", "arw", "msr", "dur"},
        # Looks like it just echoes the Arabic transliteration
        any(core.startswith(x) for x in ["say", "jia", "shea"]) and len(core) <= 6,
    ]
    return any(bad_signals)


def fix_batch(roots_batch: list) -> dict:
    """
    Given a list of root dicts, return a mapping of root_id -> {
        "rootMeaning": str,
        "babs": { bab_id: { "verbMeaning": str } }
    }
    """
    # Build prompt
    roots_info = []
    for r in roots_batch:
        bab_list = []
        for b in r["babs"]:
            bab_list.append({
                "id": b["id"],
                "form": b["romanNumeral"],
                "arabicPattern": b["arabicPattern"],
                "formMeaning": b.get("semanticMeaning") or b.get("meaning") or "",
            })
        roots_info.append({
            "id": r["id"],
            "root": r["root"],
            "currentMeaning": r["meaning"],
            "babs": bab_list,
        })

    prompt = f"""You are an expert in Arabic linguistics, specifically Quranic Arabic verb forms (أبواب الفعل).

I will give you a list of Arabic verb roots. For each root:
1. Provide the correct short English meaning for the root itself (2-5 words, starting with "to"). Fix any wrong/transliterated meanings.
2. For each bab (verb form/pattern) of that root, provide a SHORT specific English phrase (3-7 words) that captures what THIS verb actually means in THAT form — combining the root meaning with the form's semantic effect.

Examples of good bab verb meanings:
- Root "to know" Form I → "to know, to be aware"
- Root "to know" Form II (intensive/causative) → "to teach, to make known"
- Root "to know" Form IV (causative) → "to inform, to notify"
- Root "to know" Form V (reflexive of II) → "to learn, to become informed"
- Root "to glorify" Form II → "to exalt greatly, to magnify"
- Root "to believe" Form IV → "to give safety, to bring to faith"

Keep meanings concise, natural English. Do NOT use generic labels like "Base form" or "Causative".

Return ONLY valid JSON in this exact structure:
{{
  "results": [
    {{
      "id": "<root id>",
      "rootMeaning": "<corrected root meaning>",
      "babs": [
        {{ "id": "<bab id>", "verbMeaning": "<specific verb meaning in this form>" }}
      ]
    }}
  ]
}}

Here are the roots to process:
{json.dumps(roots_info, ensure_ascii=False, indent=2)}"""

    message = client.messages.create(
        model="claude-haiku-4-5-20251001",
        max_tokens=4096,
        messages=[{"role": "user", "content": prompt}],
    )

    text = message.content[0].text.strip()
    # Strip markdown code fences if present
    if text.startswith("```"):
        text = text.split("```")[1]
        if text.startswith("json"):
            text = text[4:]
    text = text.strip()

    data = json.loads(text)
    result = {}
    for item in data["results"]:
        result[item["id"]] = {
            "rootMeaning": item["rootMeaning"],
            "babs": {b["id"]: b["verbMeaning"] for b in item["babs"]},
        }
    return result


def main():
    print(f"Loading data from {DATA_PATH}")
    with open(DATA_PATH, encoding="utf-8") as f:
        data = json.load(f)

    roots = data["roots"]
    print(f"Total roots: {len(roots)}")

    # Decide which roots to process:
    # - All roots get bab verbMeaning added
    # - Only bad meanings get rootMeaning fixed
    # For efficiency, process ALL roots (to get bab meanings) but Claude will fix meanings too

    # Check if a progress checkpoint exists
    checkpoint_path = DATA_PATH.parent / "verbsData_checkpoint.json"
    processed_ids = set()
    if checkpoint_path.exists():
        print("Found checkpoint, resuming...")
        with open(checkpoint_path, encoding="utf-8") as f:
            checkpoint = json.load(f)
        # Build a set of already-processed root ids
        for r in checkpoint["roots"]:
            if r["babs"] and "verbMeaning" in r["babs"][0]:
                processed_ids.add(r["id"])
        data = checkpoint
        roots = data["roots"]
        print(f"Already processed: {len(processed_ids)} roots")

    # Build index for quick lookup
    root_index = {r["id"]: i for i, r in enumerate(roots)}

    # Process in batches
    to_process = [r for r in roots if r["id"] not in processed_ids]
    print(f"Roots to process: {len(to_process)}")

    total_batches = (len(to_process) + BATCH_SIZE - 1) // BATCH_SIZE

    for batch_num, i in enumerate(range(0, len(to_process), BATCH_SIZE)):
        batch = to_process[i : i + BATCH_SIZE]
        print(f"Batch {batch_num + 1}/{total_batches}: processing {[r['root'] for r in batch]}")

        try:
            fixes = fix_batch(batch)
        except Exception as e:
            print(f"  ERROR on batch: {e}")
            # Save checkpoint and exit
            with open(checkpoint_path, "w", encoding="utf-8") as f:
                json.dump(data, f, ensure_ascii=False, indent=2)
            print(f"Checkpoint saved. Re-run to continue.")
            sys.exit(1)

        # Apply fixes to data
        for root_id, fix in fixes.items():
            idx = root_index.get(root_id)
            if idx is None:
                continue
            r = roots[idx]

            # Fix root meaning
            new_meaning = fix.get("rootMeaning", "").strip()
            if new_meaning:
                r["meaning"] = new_meaning

            # Add verbMeaning to each bab
            bab_fixes = fix.get("babs", {})
            for bab in r["babs"]:
                vm = bab_fixes.get(bab["id"], "").strip()
                if vm:
                    bab["verbMeaning"] = vm

        # Save checkpoint every batch
        with open(checkpoint_path, "w", encoding="utf-8") as f:
            json.dump(data, f, ensure_ascii=False)

        print(f"  Done. Sleeping 0.5s...")
        time.sleep(0.5)

    # All done — write final output
    output_path = DATA_PATH
    with open(output_path, "w", encoding="utf-8") as f:
        json.dump(data, f, ensure_ascii=False)

    # Clean up checkpoint
    if checkpoint_path.exists():
        checkpoint_path.unlink()

    print(f"\nDone! Updated {DATA_PATH}")
    print("Sample check:")
    for r in roots[:3]:
        print(f"  {r['root']}: {r['meaning']}")
        for b in r["babs"][:2]:
            print(f"    Bab {b['romanNumeral']}: {b.get('verbMeaning', '(no verbMeaning)')}")


if __name__ == "__main__":
    if not os.environ.get("ANTHROPIC_API_KEY"):
        print("ERROR: Set ANTHROPIC_API_KEY environment variable first")
        sys.exit(1)
    main()

File: scripts/fix-meanings.mjs
import Anthropic from '@anthropic-ai/sdk';
import { readFileSync, writeFileSync } from 'fs';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';

const __dirname = dirname(fileURLToPath(import.meta.url));
const DATA_PATH = join(__dirname, '../src/data/verbsData.json');

const API_KEY = process.argv[2] || process.env.ANTHROPIC_API_KEY;
if (!API_KEY) { console.error('Usage: node fix-meanings.mjs <api-key>'); process.exit(1); }
const client = new Anthropic({ apiKey: API_KEY });

const data = JSON.parse(readFileSync(DATA_PATH, 'utf8'));
const roots = data.roots;

// Get sample conjugation for a root (first available arabic form that isn't '-')
function getSampleConjugation(root) {
  for (const bab of root.babs) {
    for (const tense of (bab.tenses || [])) {
      for (const conj of (tense.conjugation || [])) {
        if (conj.arabic && conj.arabic !== '-' && !conj.arabic.includes('_') && !conj.arabic.includes('#')) {
          return conj.arabic;
        }
      }
    }
  }
  return null;
}

// Get a few verse references
function getSampleRefs(root) {
  const refs = [];
  for (const bab of root.babs) {
    for (const tense of (bab.tenses || [])) {
      if (tense.references?.length) {
        refs.push(...tense.references.slice(0, 2));
        if (refs.length >= 3) break;
      }
    }
    if (refs.length >= 3) break;
  }
  return refs.slice(0, 3);
}

// Get all forms used
function getForms(root) {
  return root.babs.map(b => `Form ${b.romanNumeral}`).join(', ');
}

async function fixBatch(batch) {
  const items = batch.map(root => {
    const conj = getSampleConjugation(root);
    const refs = getSampleRefs(root);
    return {
      id: root.id,
      root: root.root,
      currentMeaning: root.meaning,
      sampleConjugation: conj || 'N/A',
      forms: getForms(root),
      refs: refs.join(', ') || 'N/A',
    };
  });

  const prompt = `You are an expert in Classical Arabic and Quranic linguistics. For each Arabic verb root below, provide the correct English meaning as used in the Quran.

Rules:
- Use infinitive form: "to X" (e.g. "to say", "to believe", "to go")
- Give the PRIMARY root meaning (Form I base meaning), not a derived form
- Use Quranic/Islamic meaning where it differs from general Arabic
- Be concise: 1-5 words max
- Also give confidence: "high", "medium", or "low"

Roots to fix:
${items.map((item, i) => `${i + 1}. Root: ${item.root} | Current (possibly wrong): "${item.currentMeaning}" | Sample conjugation: ${item.sampleConjugation} | Forms in Quran: ${item.forms} | Verse refs: ${item.refs}`).join('\n')}

Respond with ONLY a JSON array, one object per root, in the same order:
[{"id":"root_id","meaning":"to X","confidence":"high|medium|low"}, ...]`;

  const response = await client.messages.create({
    model: 'claude-sonnet-4-6',
    max_tokens: 2048,
    messages: [{ role: 'user', content: prompt }],
  });

  const text = response.content[0].text.trim();
  // Extract JSON array from response
  const jsonMatch = text.match(/\[[\s\S]*\]/);
  if (!jsonMatch) throw new Error(`No JSON array in response: ${text.slice(0, 200)}`);
  return JSON.parse(jsonMatch[0]);
}

async function main() {
  const BATCH_SIZE = 40;
  const lowConfidence = [];
  let fixed = 0;
  let errors = 0;

  console.log(`Processing ${roots.length} roots in batches of ${BATCH_SIZE}...`);

  for (let i = 0; i < roots.length; i += BATCH_SIZE) {
    const batch = roots.slice(i, i + BATCH_SIZE);
    const batchNum = Math.floor(i / BATCH_SIZE) + 1;
    const totalBatches = Math.ceil(roots.length / BATCH_SIZE);
    process.stdout.write(`Batch ${batchNum}/${totalBatches}... `);

    try {
      const results = await fixBatch(batch);

      for (const result of results) {
        const root = roots.find(r => r.id === result.id);
        if (!root) continue;

        const oldMeaning = root.meaning;
        root.meaning = result.meaning;
        fixed++;

        if (result.confidence !== 'high') {
          lowConfidence.push({ id: result.id, root: root.root, oldMeaning, newMeaning: result.meaning, confidence: result.confidence });
        }
      }

      console.log(`done (${results.length} roots)`);
    } catch (err) {
      console.log(`ERROR: ${err.message}`);
      errors++;
    }

    // Small delay to avoid rate limiting
    if (i + BATCH_SIZE < roots.length) {
      await new Promise(r => setTimeout(r, 500));
    }
  }

  // Save updated data
  writeFileSync(DATA_PATH, JSON.stringify(data, null, 2), 'utf8');
  console.log(`\nDone! Fixed ${fixed} roots, ${errors} batch errors.`);

  if (lowConfidence.length > 0) {
    console.log(`\n⚠️  ${lowConfidence.length} low/medium confidence entries to review:`);
    lowConfidence.forEach(r => {
      console.log(`  ${r.root} (${r.id}): "${r.oldMeaning}" → "${r.newMeaning}" [${r.confidence}]`);
    });

    // Save review list
    const reviewPath = join(__dirname, 'meanings-to-review.json');
    writeFileSync(reviewPath, JSON.stringify(lowConfidence, null, 2), 'utf8');
    console.log(`\nReview list saved to scripts/meanings-to-review.json`);
  }
}

main().catch(console.error);
